include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -993,18 +1003,28 @@ struct __parallel_find_backward_tag
     using _AtomicType = oneapi::dpl::__internal::__difference_t<_RangeType>;
 #endif
 
+    using FoundLocalReduceOp = __dpl_sycl::__maximum<_AtomicType>;
+
     template <typename _DiffType>
     constexpr static _AtomicType __init_value(_DiffType)
     {
         return _AtomicType{-1};
     }
 
-    template <typename _TAtomic>
+    // As far as we make search from end to the begin of data, we should save the last (maximal) found state in the __save_state_to methods.
+
+    template <sycl::access::address_space _Space>
+    static void
+    __save_state_to(__dpl_sycl::__atomic_ref<_AtomicType, _Space>& __atomic, _AtomicType __new_state)
+    {
+        __atomic.fetch_max(__new_state);
+    }
+
+    template <typename _TFoundState>
     static void
-    __save_state_to_atomic(_TAtomic& __found, _AtomicType __new_state)
+    __save_state_to(_TFoundState& __found, _AtomicType __new_state)
```

Comment:
I'd prefer having these functions be named `__save_state_to` and `__save_state_to_atomic`, rather than have them be overloads of the same function. 
We should make it clear for the maintainer when atomic operations are occurring and not make them decipher it by looking at argument types, etc.